### PR TITLE
 add news rec methods NAML NRMS NPA LSTUR

### PR DESCRIPTION
**Pull request by yjw1029**
### Description

Add news rec methods (NAML NRMS NPA LSTUR)

### Related Issues

#1079 

### Checklist:

- [x] I have followed the [contribution guidelines](../CONTRIBUTING.md) and code style for this project.
- [x] I have added tests covering my contributions.
- [x] I have updated the documentation accordingly.
- [x] This PR is being made to `staging` and not `master`.

### CODE REVIEW

File: reco_utils/recommender/deeprec/deeprec_utils.py [hunk 1]
```diff
@@ -553,16 +554,26 @@ def cal_metric(labels, preds, metrics):
...
             res["mean_mrr"] = round(mean_mrr, 4)
         elif metric.startswith("ndcg"):  # format like:  ndcg@2;4;6;8
             ndcg_list = [1, 2]
             ks = metric.split('@')
             if len(ks) > 1:
                 ndcg_list = [int(token) for token in ks[1].split(';')]
             for k in ndcg_list:
-                ndcg_temp = np.mean(ndcg_score(labels, preds, k))
-                res["ndcg@{0}".format(k)] = round(ndcg_temp, 4)
+                ndcg_temp = np.mean(
```
Comment (yueguoguo) [inline reco_utils/recommender/deeprec/deeprec_utils.py:570]: Why the change of this? `labels` and `preds` themselves are lists so `ndcg_score` directly calculate the metrics for them?

File: reco_utils/recommender/newsrec/IO/iterator.py [hunk 2]
```diff
@@ -0,0 +1,24 @@
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License.
+
+import numpy as np
+import tensorflow as tf
+import abc
+
+
+class BaseIterator(object):
```
Comment (yueguoguo) [inline reco_utils/recommender/newsrec/IO/iterator.py:9]: Can we reuse the iterator from `deeprec`?

File: reco_utils/recommender/newsrec/models/base_model.py [hunk 3]
```diff
@@ -0,0 +1,289 @@
+# Copyright (c) Microsoft Corporation. All rights reserved.
```
Comment (yueguoguo) [inline reco_utils/recommender/newsrec/models/base_model.py:1]: Similar to the above, can we reuse the `BaseModel` object from `deeprec`. I see the two object have similar interfaces and implementation.

File: README.md [hunk 4]
```diff
@@ -56,7 +56,11 @@ The table below lists the recommender algorithms currently available in the repo
 | Factorization Machine (FM) / Field-Aware FM (FFM) | [Python CPU](notebooks/02_model/fm_deep_dive.ipynb) | Content-Based Filtering | Algorithm that predict labels with user/item features |
 | FastAI Embedding Dot Bias (FAST) | [Python CPU / Python GPU](notebooks/00_quick_start/fastai_movielens.ipynb) | Collaborative Filtering | General purpose algorithm with embeddings and biases for users and items |
 | LightGBM/Gradient Boosting Tree<sup>*</sup> | [Python CPU](notebooks/00_quick_start/lightgbm_tinycriteo.ipynb) / [PySpark](notebooks/02_model/mmlspark_lightgbm_criteo.ipynb) | Content-Based Filtering | Gradient Boosting Tree algorithm for fast training and low memory usage in content-based problems |
+| Neural News Recommendation with Long- and Short-term User Representations (LSTUR) | [Python CPU / Python GPU](notebooks/00_quick_start/lstur_synthetic.ipynb) | Content-Based Filtering | Deep learning algorithm considering users' long- and short-term preference to provide powerful news or article recommendations |
```
[the remark this answers is not in the document]
Comment (yjw1029) [reply README.md:59]: I just write the same as DKN. Of course, you can remove it😁.

File: reco_utils/recommender/deeprec/deeprec_utils.py (hunk 1 above)
Comment (yjw1029) [reply reco_utils/recommender/deeprec/deeprec_utils.py:570]: I communicate with Jianxun, and make this change. In the old version, the labels and preds are 2-dim arrays, so every input label array and pred array of different impressions should have items of the same numbers. Current implementation support to compute mrr_scores/ndcg scores of different impressions that have different items. The outputs of both versions are mean values of mrr/ndcg scores.

File: reco_utils/recommender/newsrec/IO/iterator.py (hunk 2 above)
Comment (yjw1029) [reply reco_utils/recommender/newsrec/IO/iterator.py:9]: Yes, I will make a PR.

File: reco_utils/recommender/newsrec/models/base_model.py (hunk 3 above)
Comment (yjw1029) [reply reco_utils/recommender/newsrec/models/base_model.py:1]: Will be include in the PR as well.

Comment (yjw1029) [reply reco_utils/recommender/newsrec/models/base_model.py:1]: BaseModel have several different parts from deeprec. They are changed to fit tf.keras instead of tf1.0.

**Comment by miguelgfierro** [comment]
> Btw, the tests on Spark environment failed and I believe there is no Spark related code submitted in the PR, is that correct?

Right now the test machines are down because of the measurements with covid in all MS. We will need to run the tests ourselves. I will add more folks to comment in the review

File: notebooks/00_quick_start/lstur_synthetic.ipynb [hunk 5]
```diff
@@ -0,0 +1,316 @@
+{
```
Comment (miguelgfierro) [inline notebooks/00_quick_start/lstur_synthetic.ipynb:1]: <p>could we please change all the <code>import *</code>  to import the actual functions that we are using in the notebook? In the rest of the notebooks we are trying to follow that rule</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/microsoft/recommenders/pull/1080/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='3'/>

File: notebooks/00_quick_start/naml_synthetic.ipynb [hunk 6]
```diff
@@ -0,0 +1,311 @@
+{
```
Comment (miguelgfierro) [inline notebooks/00_quick_start/naml_synthetic.ipynb:1]: <p>same comment here with the imports</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/microsoft/recommenders/pull/1080/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='3'/>

Comment (miguelgfierro) [inline notebooks/00_quick_start/naml_synthetic.ipynb:1]: <p>it would be great to add also the integration tests here. Jianxun added some for the SLiRec notebook, see here: <a href="https://github.com/microsoft/recommenders/blob/master/tests/integration/test_notebooks_gpu.py#L239" target="_blank">https://github.com/microsoft/recommenders/blob/master/tests/integration/test_notebooks_gpu.py#L239</a></p><p> </p><p>See instructions here <a href="https://github.com/microsoft/recommenders/tree/master/tests#developing-smoke-and-integration-tests-with-papermill" target="_blank">https://github.com/microsoft/recommenders/tree/master/tests#developing-smoke-and-integration-tests-with-papermill</a></p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/microsoft/recommenders/pull/1080/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='14'/>

File: notebooks/00_quick_start/npa_synthetic.ipynb [hunk 7]
```diff
@@ -0,0 +1,298 @@
+{
```
Comment (miguelgfierro) [inline notebooks/00_quick_start/npa_synthetic.ipynb:1]: <p>same comments for imports and tests for all notebooks</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/microsoft/recommenders/pull/1080/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='0'/>

File: reco_utils/recommender/newsrec/models/lstur.py [hunk 8]
```diff
@@ -0,0 +1,160 @@
...
+        user_embedding_layer = layers.Embedding(hparams.user_num,
+                                 hparams.gru_unit,
+                                 trainable=True,
+                                 embeddings_initializer='zeros')
+        
+        long_u_emb = layers.Reshape((hparams.gru_unit,))(user_embedding_layer(user_indexes))
+        click_title_presents = layers.TimeDistributed(titleencoder)(his_input_title)
+
+        if type=='ini':
+            user_present = layers.GRU(hparams.gru_unit,)(
```
Comment (miguelgfierro) [inline reco_utils/recommender/newsrec/models/lstur.py:83]: I have a curiosity, in many occasions I've seen that you have used GRU instead of LSTM. Could you please explain the reason?

File: reco_utils/recommender/newsrec/models/naml.py [hunk 9]
```diff
@@ -0,0 +1,272 @@
...
+        Returns:
+            obj: a model used to train.
+            obj: a model used to evaluate and inference.
+        """
+
+        model, scorer = self._build_naml()
+        return model, scorer
+
+    def _build_userencoder(self, newsencoder):
+        '''The main function to create user encoder of NAML.
```
Comment (miguelgfierro) [inline reco_utils/recommender/newsrec/models/naml.py:63]: I've noticed that in some files you added `'''` and in others `"""`. Normally what we do is to use black, that automatically formats everything when you save. I can do it for you if you want. 

In case you want to add it yourself, I can tell you very quickly how to activate it in VSCode. You just need to have a conda environment with black installed and then add this line to your configuration https://github.com/miguelgfierro/codebase/blob/master/minsc/vscode_settings.json#L7

If you or anyone in your team want to know how to activate this, I'll be happy to jump on a quick call to help you do this very quickly

File: reco_utils/recommender/newsrec/models/naml.py [hunk 10]
```diff
@@ -0,0 +1,272 @@
...
+        '''build title encoder of NAML news encoder.
+
+        Args:
+            embedding_layer(obj): a word embedding layer.
+        
+        Return:
+            obj: the title encoder of NAML.
+        '''
+        hparams = self.hparams
+        sequences_input_title = keras.Input(shape=(hparams.title_size,),dtype='int32')
```
Comment (miguelgfierro) [inline reco_utils/recommender/newsrec/models/naml.py:120]: I noticed that now you are using keras instead of TF. Has the team moved to keras?

File: README.md [hunk 11]
```diff
@@ -56,7 +56,11 @@ The table below lists the recommender algorithms currently available in the repo
...
+| Neural News Recommendation with Attentive Multi-View Learning (NAML) | [Python CPU / Python GPU](notebooks/00_quick_start/naml_synthetic.ipynb) | Content-Based Filtering | Deep learning news recommendation algorithm with multi-view learning |
 | Neural Collaborative Filtering (NCF) | [Python CPU / Python GPU](notebooks/00_quick_start/ncf_movielens.ipynb) | Collaborative Filtering | Deep learning algorithm with enhanced performance for implicit feedback |
+| Neural News Recommendation with Personalized Attention (NPA) | [Python CPU / Python GPU](notebooks/00_quick_start/npa_synthetic.ipynb) | Content-Based Filtering | Deep learning news recommendation algorithm ultilizing personalized attention |
+| Neural News Recommendation with Multi-Head Self-Attention (NRMS) | [Python CPU / Python GPU](notebooks/00_quick_start/nrms_synthetic.ipynbb) | Content-Based Filtering | Deep learning news recommendation algorithm ultilizing multi-head self-attention |
```
Comment (gramhagen) [inline README.md:63]: The rest of the algorithms in this table are not domain specific. Is it important to have the news focus here? It may lead to people overlooking it as an option for other domains.

File: reco_utils/recommender/newsrec/IO/news_iterator.py [hunk 12]
```diff
@@ -0,0 +1,308 @@
...
+        user_indexes = []
+        candidate_news_indexes = []
+        click_news_indexes = []
+        cnt = 0
+
+        with tf.gfile.GFile(infile, "r") as rd:
+            while True:
+                line = rd.readline()
+                if not line:
+                    break
```
Comment (gramhagen) [inline reco_utils/recommender/newsrec/IO/news_iterator.py:103]:
```suggestion
            for line in rd:
```

File: reco_utils/recommender/newsrec/IO/news_iterator.py [hunk 13]
```diff
@@ -0,0 +1,308 @@
...
+        imp_indexes = np.asarray(imp_indexes, dtype=np.int32)
+        user_indexes = np.asarray(user_indexes, dtype=np.int32)
+        candidate_news_index_batch = np.asarray(
+            candidate_news_indexes, dtype=np.int64
+        )
+        click_news_index_batch = np.asarray(click_news_indexes, dtype=np.int64)
+        return [imp_indexes, user_indexes, click_news_index_batch, candidate_news_index_batch], labels
+
+
+class NewsTestIterator(BaseIterator):
```
Comment (gramhagen) [inline reco_utils/recommender/newsrec/IO/news_iterator.py:159]: is this the same as NewsTrainIterator with npratio=0? Do we need separate classes? Or can we extend that class to reduce redundant code?

File: reco_utils/recommender/newsrec/IO/news_iterator.py [hunk 14]
```diff
@@ -0,0 +1,308 @@
...
+        
+        Args:
+            label_list (list): a list of ground-truth labels.
+            imp_indexes (list): a list of impression indexes.
+            user_indexes (list): a list of user indexes.
+            candidate_news_indexes (list): the candidate news article's words indices
+            click_news_indexes (list): words indices for user's clicked news articles
+            
+        Returns:
+            dict: A dictionary, contains multiple numpy arrays that are convenient for further operation.
```
Comment (gramhagen) [inline reco_utils/recommender/newsrec/IO/news_iterator.py:298]: the return value is a list, should it be a dict instead? that would certainly make it easier to extract results without needing to know the order of the output.

File: reco_utils/recommender/newsrec/models/base_model.py [hunk 15]
```diff
@@ -0,0 +1,289 @@
...
+import numpy as np
+import tensorflow as tf
+from tensorflow import keras
+from reco_utils.recommender.deeprec.deeprec_utils import cal_metric
+
+
+__all__ = ["BaseModel"]
+
+
+class BaseModel:
```
Comment (gramhagen) [inline reco_utils/recommender/newsrec/models/base_model.py:16]: better to use new style class here since we're still support python 2
```suggestion
class BaseModel(object):
```

File: reco_utils/recommender/newsrec/newsrec_utils.py [hunk 16]
```diff
@@ -0,0 +1,291 @@
...
+        title_size=flags["title_size"] if "title_size" in flags else None,
+        body_size=flags["body_size"] if "body_size" in flags else None,
+        word_emb_dim=flags["word_emb_dim"] if "word_emb_dim" in flags else None,
+        word_size=flags["word_size"] if "word_size" in flags else None,
+        user_num=flags["user_num"] if "user_num" in flags else None,
+        vert_num=flags["vert_num"] if "vert_num" in flags else None,
+        subvert_num=flags["subvert_num"] if "subvert_num" in flags else None,
+        his_size=flags["his_size"] if "his_size" in flags else None,
+        npratio=flags["npratio"] if "npratio" in flags else None,
+        dropout=flags["dropout"] if "dropout" in flags else 0.0,
```
Comment (gramhagen) [inline reco_utils/recommender/newsrec/newsrec_utils.py:242]: This could all be a bit cleaner using get, e.g.
```suggestion
        npratio=flags.get("npratio"),
        dropout=flags.get("dropout", 0.0),
```

File: reco_utils/recommender/newsrec/newsrec_utils.py [hunk 17]
```diff
@@ -0,0 +1,291 @@
...
+    """
+    if yaml_file is not None:
+        config = load_yaml(yaml_file)
+        config = flat_config(config)
+    else:
+        config = {}
+
+    if kwargs:
+        for name, value in six.iteritems(kwargs):
+            config[name] = value
```
Comment (gramhagen) [inline reco_utils/recommender/newsrec/newsrec_utils.py:288]:
```suggestion
    config.update(kwargs)
```

File: tests/smoke/test_newsrec_model.py [hunk 18]
```diff
@@ -0,0 +1,134 @@
...
+from reco_utils.recommender.newsrec.models.base_model import BaseModel
+from reco_utils.recommender.newsrec.models.nrms import NRMSModel
+from reco_utils.recommender.newsrec.models.naml import NAMLModel
+from reco_utils.recommender.newsrec.models.lstur import LSTURModel
+from reco_utils.recommender.newsrec.models.npa import NPAModel
+from reco_utils.recommender.newsrec.IO.news_iterator import NewsTrainIterator, NewsTestIterator
+from reco_utils.recommender.newsrec.IO.naml_iterator import NAMLTrainIterator, NAMLTestIterator
+
+
+@pytest.fixture
```
Comment (gramhagen) [inline tests/smoke/test_newsrec_model.py:18]: it would be better to use a temporary directory so that test resources do not accumulate on any build machines. https://docs.pytest.org/en/latest/tmpdir.html

File: reco_utils/recommender/newsrec/models/base_model.py (hunk 15 above)
Comment (miguelgfierro) [reply reco_utils/recommender/newsrec/models/base_model.py:16]: afaik we are not supporting python 2:
- https://github.com/microsoft/recommenders/blob/master/scripts/generate_conda_file.py#L38
- https://github.com/microsoft/recommenders/blob/master/setup.py#L42
is there any place where we have python 2 stuff that I missed?

File: tests/smoke/test_newsrec_model.py (hunk 18 above)
Comment (miguelgfierro) [reply tests/smoke/test_newsrec_model.py:18]: good suggestion, we have a fixture that generates a tmp file that is removed when the tests are finished: https://github.com/microsoft/recommenders/blob/master/tests/conftest.py#L23, just need to add `tmp` in the function for it to work, like here https://github.com/microsoft/recommenders/blob/master/tests/integration/test_movielens.py#L69 (there is no need to import because it is a fixture)

File: reco_utils/recommender/newsrec/models/base_model.py (hunk 15 above)
Comment (gramhagen) [reply reco_utils/recommender/newsrec/models/base_model.py:16]: Ah. That's great. Nevermind then.

File: reco_utils/recommender/newsrec/models/layers.py [hunk 19]
```diff
@@ -0,0 +1,280 @@
...
+
+        attention = K.squeeze(attention, axis=2)
+        attention = K.exp(attention)
+        attention_weight = attention / (K.sum(attention, axis=-1, keepdims=True) + K.epsilon())
+
+        attention_weight = K.expand_dims(attention_weight)
+        weighted_input = inputs * attention_weight
+        return K.sum(weighted_input, axis=1)
+
+    def compute_mask(self, input, input_mask=None):
```
[the remark this answers is not in the document]
Comment (yjw1029) [reply reco_utils/recommender/newsrec/models/layers.py:67]: It means the layer will not output mask value to next layers. I will add docstring to this method.

File: reco_utils/recommender/newsrec/models/lstur.py (hunk 8 above)
Comment (yjw1029) [reply reco_utils/recommender/newsrec/models/lstur.py:83]: GRU is faster and smaller than LSTM. And in our experiment, GRU is as effective as LSTM.

File: reco_utils/recommender/newsrec/models/naml.py (hunk 9 above)
Comment (yjw1029) [reply reco_utils/recommender/newsrec/models/naml.py:63]: Thanks, I will try add it.

File: reco_utils/recommender/newsrec/models/naml.py (hunk 10 above)
Comment (yjw1029) [reply reco_utils/recommender/newsrec/models/naml.py:120]: Actually I use tf.keras. Different from Jianxun's team, the code of our(Fangzhao's) team is implemented by keras.

File: reco_utils/recommender/newsrec/IO/news_iterator.py (hunk 13 above)
Comment (yjw1029) [reply reco_utils/recommender/newsrec/IO/news_iterator.py:159]: Thanks for the suggestions. I will try to change the code and combine these two iterators.

File: notebooks/00_quick_start/naml_synthetic.ipynb (hunk 6 above)
Comment (anargyri) [inline notebooks/00_quick_start/naml_synthetic.ipynb:1]: <p>Could you explain what "vert" and "subvert" mean here?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/microsoft/recommenders/pull/1080/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='1'/>

File: reco_utils/recommender/newsrec/IO/iterator.py (hunk 2 above)
Comment (yueguoguo) [reply reco_utils/recommender/newsrec/IO/iterator.py]: Just directly merge the new PR to this one. We don't need to create a separate one.

File: reco_utils/recommender/newsrec/models/base_model.py (hunk 3 above)
Comment (yueguoguo) [reply reco_utils/recommender/newsrec/models/base_model.py:1]: Is there particular reason why the change? It might be great to make the interface consistent across different algorithms.

File: reco_utils/recommender/newsrec/models/naml.py (hunk 10 above)
Comment (miguelgfierro) [reply reco_utils/recommender/newsrec/models/naml.py:120]: got it thanks

File: notebooks/00_quick_start/naml_synthetic.ipynb (hunk 6 above)
Comment (yjw1029) [reply notebooks/00_quick_start/naml_synthetic.ipynb:1]: It means news vertical and subvertical.

File: notebooks/00_quick_start/README.md [hunk 20]
```diff
@@ -11,7 +11,11 @@ data preparation, model building, and model evaluation by using the utility func
...
 | [lightgbm](lightgbm_tinycriteo.ipynb) | Criteo | Python CPU | Utilizing LightGBM Boosting Tree to predict whether or not a user has clicked on an e-commerce ad |
+| [lstur](lstur_synthetic.ipynb) | Synthetic Data | Python CPU, GPU | Utilizing the Neural News Recommendation with Long- and Short-term User Representations (LSTUR) [9] for news recommendation, in a Python+GPU (Tensorflow) enviroment.
+| [naml](naml_synthetic.ipynb) | Synthetic Data | Python CPU, GPU | Utilizing the Neural News Recommendation with Attentive Multi-View Learning (NAML) [7] to algorithm for news recommendation using news verticle, subverticle, title and body information, in a Python+GPU (Tensroflow) environment.  
 | [ncf](ncf_movielens.ipynb) | MovieLens | Python CPU, GPU |  Utilizing Neural Collaborative Filtering (NCF) [1] to predict movie ratings in a Python+GPU (TensorFlow) environment.
+| [npa](npa_synthetic.ipynb) | Synthetic Data | Python CPU, GPU | Utilizing the Neural News Recommendation with Personalized Attention (NPA) [10] for news recommendation, in a Python+GPU (Tensroflow) environment.
```
Comment (miguelgfierro) [inline notebooks/00_quick_start/README.md:17]: minor typo: `Tensorflow` instead of `Tensroflow`

File: notebooks/00_quick_start/lstur_synthetic.ipynb (hunk 5 above)
Comment (anargyri) [inline notebooks/00_quick_start/lstur_synthetic.ipynb:1]: <p>cpatures&nbsp;= captures</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/microsoft/recommenders/pull/1080/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='1'/>

File: reco_utils/recommender/newsrec/models/base_model.py (hunk 3 above)
Comment (yjw1029) [reply reco_utils/recommender/newsrec/models/base_model.py:1]: It seems tf.keras can combine with tf lower level API. I will try to merge newsrec into deeprec.

Comment (yueguoguo) [reply reco_utils/recommender/newsrec/models/base_model.py:1]: Knowing the difference of APIs may lead to some discrepancies in evaluation metrics, so the merging may not make sense.

File: README.md [hunk 21]
```diff
@@ -56,7 +56,11 @@ The table below lists the recommender algorithms currently available in the repo
 | Factorization Machine (FM) / Field-Aware FM (FFM) | [Python CPU](notebooks/02_model/fm_deep_dive.ipynb) | Content-Based Filtering | Algorithm that predict labels with user/item features |
 | FastAI Embedding Dot Bias (FAST) | [Python CPU / Python GPU](notebooks/00_quick_start/fastai_movielens.ipynb) | Collaborative Filtering | General purpose algorithm with embeddings and biases for users and items |
 | LightGBM/Gradient Boosting Tree<sup>*</sup> | [Python CPU](notebooks/00_quick_start/lightgbm_tinycriteo.ipynb) / [PySpark](notebooks/02_model/mmlspark_lightgbm_criteo.ipynb) | Content-Based Filtering | Gradient Boosting Tree algorithm for fast training and low memory usage in content-based problems |
+| Neural Recommendation with Long- and Short-term User Representations (LSTUR) | [Python CPU / Python GPU](notebooks/00_quick_start/lstur_synthetic.ipynb) | Content-Based Filtering | Neural recommendation algorithm with long- and short-term user interest modeling |
```
Comment (miguelgfierro) [inline README.md:59]: minor comment, we are adding the mark `<sup>*</sup>` when the algo was designed by MS

File: tests/unit/test_newsrec_model.py [hunk 22]
```diff
@@ -0,0 +1,116 @@
...
+@pytest.fixture
+def resource_path():
+    return os.path.dirname(os.path.realpath(__file__))
+
+@pytest.mark.gpu
+@pytest.mark.newsrec
+@pytest.mark.nrms
+def test_nrms_component_definition(tmp):
+    data_path = tmp
+    yaml_file = os.path.join(data_path, 'nrms.yaml')
```
Comment (miguelgfierro) [inline tests/unit/test_newsrec_model.py:25]: minor comment, it is possible to use directly tmp:
```
yaml_file = os.path.join(tmp, 'nrms.yaml')
wordEmb_file = os.path.join(tmp, 'embedding.npy')
```

File: tests/smoke/test_newsrec_model.py [hunk 23]
```diff
@@ -0,0 +1,125 @@
...
+    iterator = NewsIterator
+    model = LSTURModel(hparams, iterator)
+
+    assert model.run_eval(valid_file) is not None
+    assert isinstance(model.fit(train_file, valid_file), BaseModel)
+
+
+@pytest.mark.smoke
+@pytest.mark.gpu
+@pytest.mark.newsrec
```
Comment (miguelgfierro) [inline tests/smoke/test_newsrec_model.py:102]: I would kindly ask to remove the markers that are not core to our libraries if it doesn't cause you a lot of trouble.

In our tests pipeline we use `@pytest.mark.gpu`, `@pytest.mark.smoke`, `@pytest.mark.integration`, etc.

Everything related to the news algo could go out: `@pytest.mark.newsrec`, `@pytest.mark.lstur`, `@pytest.mark.npa`, etc

File: tests/unit/test_newsrec_utils.py [hunk 24]
```diff
@@ -0,0 +1,145 @@
...
+        assert res["impression_index_batch"].shape == (hparams.batch_size, 1)
+        assert res["user_index_batch"].shape == (hparams.batch_size, 1)
+        assert res["clicked_news_batch"].shape == (hparams.batch_size, hparams.his_size, hparams.doc_size)
+        assert res["candidate_news_batch"].shape == (hparams.batch_size, 1, hparams.doc_size)
+
+
+
+@pytest.mark.gpu
+@pytest.mark.newsrec
+def test_NAML_iterator(tmp):
```
Comment (miguelgfierro) [inline tests/unit/test_newsrec_utils.py:97]: would it be possible to follow a non-capitalize format for names? `test_NAML_iterator` would be `test_naml_iterator`

File: tests/unit/test_newsrec_utils.py [hunk 25]
```diff
@@ -0,0 +1,145 @@
...
+        assert isinstance(res, dict)
+        assert len(res) == 5
+        assert res["impression_index_batch"].shape == (hparams.batch_size, 1)
+        assert res["user_index_batch"].shape == (hparams.batch_size, 1)
+        assert res["clicked_news_batch"].shape == (hparams.batch_size, hparams.his_size, hparams.doc_size)
+        assert res["candidate_news_batch"].shape == (hparams.batch_size, 1, hparams.doc_size)
+
+
+
+@pytest.mark.gpu
```
Comment (miguelgfierro) [inline tests/unit/test_newsrec_utils.py:95]: I run the tests and found that this specific test takes 65.88s. Some questions:
1) do you know where is the bottleneck?
2) is it possible to reduce the time?

**Comment by miguelgfierro** [comment]
hey @yjw1029 @yueguoguo @gramhagen @anargyri, just FYI, I run all the tests in Prometheus, new and old ones and everything passes

File: tests/unit/test_newsrec_utils.py (hunk 25 above)
Comment (yjw1029) [reply tests/unit/test_newsrec_utils.py:82]: 1. It may because naml data file contains news body, which is much longer than news title. 
2. I remove the iterator loop to reduce the time.

**Comment by yjw1029** [comment]
hi @miguelgfierro. Thanks for your comment. I change the problems you mentioned above😁. Are there any issues in the code?

**Comment by miguelgfierro** [comment]
@yjw1029 from my part this is ready to go. 

@gramhagen is there anything else that you would like to review?

**Comment by miguelgfierro** [comment]
after speaking with @gramhagen, I'll merge this.

@yjw1029 amazing work
